arangod/Replication2/StateMachines/Prototype/PrototypeStateMachine.cpp

```diff
@@ -161,29 +204,63 @@ auto PrototypeLeaderState::get(std::string key) -> std::optional<std::string> {
       });
 }
 
-auto PrototypeLeaderState::getSnapshot()
-    -> ResultT<std::unordered_map<std::string, std::string>> {
-  return guardedData.doUnderLock(
-      [](auto& core) -> ResultT<std::unordered_map<std::string, std::string>> {
-        if (!core) {
-          return ResultT<std::unordered_map<std::string, std::string>>::error(
-              TRI_ERROR_CLUSTER_NOT_LEADER);
-        }
-        std::unordered_map<std::string, std::string> result{core->store.begin(),
-                                                            core->store.end()};
-        return result;
-      });
+auto PrototypeLeaderState::getSnapshot(LogIndex waitForIndex)
+    -> futures::Future<ResultT<std::unordered_map<std::string, std::string>>> {
+  return guardedData.doUnderLock([waitForIndex,
+                                  self = shared_from_this()](auto& core)
+                                     -> futures::Future<
+                                         ResultT<std::unordered_map<
+                                             std::string, std::string>>> {
+    if (!core) {
+      return ResultT<std::unordered_map<std::string, std::string>>::error(
+          TRI_ERROR_CLUSTER_NOT_LEADER);
+    }
+    // TODO use then
+    return core->waitForApplied(waitForIndex)
+        .thenValue([&core, self](auto&& r)
+                       -> ResultT<
+                           std::unordered_map<std::string, std::string>> {
+          if (!core) {
+            return ResultT<std::unordered_map<std::string, std::string>>::error(
+                TRI_ERROR_CLUSTER_NOT_LEADER);
+          }
+          std::unordered_map<std::string, std::string> result{
+              core->store.begin(), core->store.end()};
+          return ResultT<std::unordered_map<std::string, std::string>>::success(
+              result);
+        });
+  });
 }
 
 PrototypeFollowerState::PrototypeFollowerState(
-    std::unique_ptr<PrototypeCore> core)
-    : guardedData(std::move(core)) {}
+    std::unique_ptr<PrototypeCore> core,
+    std::shared_ptr<IPrototypeNetworkInterface> networkInterface)
+    : logIdentifier(core->logId),
+      guardedData(std::move(core)),
+      networkInterface(std::move(networkInterface)) {}
 
 auto PrototypeFollowerState::acquireSnapshot(ParticipantId const& destination,
-                                             LogIndex) noexcept
+                                             LogIndex waitForIndex) noexcept
     -> futures::Future<Result> {
-  // TODO
-  return {TRI_ERROR_NO_ERROR};
+  ResultT<std::shared_ptr<IPrototypeLeaderInterface>> leader =
+      networkInterface->getLeaderInterface(destination);
+  if (leader.fail()) {
+    return leader.result();
+  }
+  return leader.get()
+      ->getSnapshot(logIdentifier, waitForIndex)
+      .thenValue([self = shared_from_this()](auto&& result) -> Result {
+        if (result.fail()) {
+          return result.result();
+        }
+
+        auto map = result.get();
```

Comment:
This will do a copy, which can be huge.
```suggestion
        auto& map = result.get();
```

Reply:
Or move the map out of the result.